cli: lift the SQL shell into its own sub-package `clisqlshell` (7/)

knz opened the pull request:
First 24 commits are from #67461 and prior.
Please ignore those (they will disappear once the prefix PR is merged)

Only the last 4 commits belong to this PR.

This is part of a family of PRs that aim to encapsulate CLI functionality into finer grained packages.

knz committed:
cli: move SQL client code to `clisqlclient`, new package `clierror`

This commit leaves the code in an intermediate stage, where the unit
tests pass but maybe the linter is unhappy. This will be cleaned up in
the next commit.

Release note: None

knz committed:
clisqlclient: rename some files

Release note: None

knz committed:
clisqlclient: move InitialSQLConnectionError to own file

Release note: None

knz committed:
cli, clisqlclient: only uses public APIs

Release note: None

knz committed:
cli: properly handle conn close errors

Release note: None

knz committed:
clisqlclient: split code in smaller files

Release note: None

knz committed:
clisqlclient: `conn_util` -> `conn`

Release note: None

knz committed:
clisqlclient: avoid depending on formatting code in conn.go

Release note: None

knz committed:
clisqlclient: move exec/formatting code to new package `clisqlexec`

Release note: None

knz committed:
cli: untangle `cliCtx` and `sqlCtx`

In a next step we'll want to lift configuration fields from `cliCtx`
and `sqlCtx` into appropriate sub-packages. Before we can do this
however, we must ensure that the fields relevant to each sub-package
are in the right structs:

- we want a `sqlConnCtx` which is scoped to establishing SQL client
  connections not just the SQL interactive shell (which has `sqlCtx`).
- we want a `sqlExecCtx` which is scoped to executing and rendering
  SQL query results on the screen, even for non-interactive use
  e.g. in `cockroach node ls`.

Here's a summary of the homing:

| Field                          | Config       | Note            |
|--------------------------------|--------------|-----------------|
| `isInteractive`                | `cliCtx`     | unchanged       |
| `echo`                         | `sqlConnCtx` | was in `sqlCtx` |
| `debugMode`                    | `sqlConnCtx` | was in `sqlCtx` |
| `embeddedMode`                 | `sqlConnCtx` | was in `sqlCtx` |
| `enableServerExecutionTimings` | `sqlConnCtx` | was in `sqlCtx` |
| `terminalOutput`               | `sqlExecCtx` | was in `cliCtx` |
| `tableDisplayFormat`           | `sqlExecCtx` | was in `cliCtx` |
| `tableBorderMode`              | `sqlExecCtx` | was in `cliCtx` |
| `showTimes`                    | `sqlExecCtx` | was in `sqlCtx` |
| `verboseTimings`               | `sqlExecCtx` | was in `sqlCtx` |

Additionally, this patch breaks the embedding of the `cliContext` by
`sqlContext` so that the lifting into a separate package remains
painless.

Release note: None

knz committed:
cli: lift config vars into sub-packages

This lifts the `cliCtx.isInteractive`, `sqlConnCtx` and `sqlExecCtx`
into sub-packages.

Release note: None

knz committed:
cli, clisqlclient: connect the config structs together

This switches the APIs from passing configs as arguments to making the
functions methods of their relevant config structs.

Release note: None

knz committed:
clisqlclient: move the handleCopyError special case to MakeQuery

We'd like to move `RunQueryAndFormatResults` to a separate package
and for this it needs to cease knowing about the internals of a
`sqlConn`.

To achieve this, this patch moves the `handleCopyError` code under
control of `MakeQuery`. Since `RunQueryAndFormatResults` is always
used in combination with `MakeQuery` anyway, this results in no change
in functionality.

Release note: None

knz committed:
clisqlclient: disentangle `RunQueryAndFormatResults`

We want to move `RunQueryAndFormatResults` to a different package and
for this it needs to cease knowing about the `sqlConn` type.

This commit does just that.

Release note: None

knz committed:
clisqlexec,clisqlclient: remove stray uses of `fmt.Errorf`

Release note: None

knz committed:
clisqlclient,clisqlexec: stop using stdout/stderr implicitly

This commit makes the output streams explicit in the API calls.

Release note: None

knz committed:
cli: define a new interface `DemoCluster`

This detaches the SQL shell from the internals of the demo cluster
implementation.

Release note: None

knz committed:
cli: move the demo cluster's tests to their own file

Release note: None

knz committed:
cli: move the demo cluster logic to own package `democluster`

The main functionality in `demo_cluster.go` was simply moved to
`democluster`.
To achieve this, some untangling was needed:

- the remaining `transientCluster` methods useful to the SQL shell are
  now all announced via an
  interface, so that the SQL shell doesn't need to know about the
  concrete type any more.

- the error reporting function `checkAndShoutError()` is moved to the
  `clierror` package, where arguably it should have been from the start.

- the `createTestCerts` function is moved to the `securitytest`
  package, where it belongs anyway.

- the `demoLocalityList` type is moved from `cli/flags_util.go` to a new
  file `democluster/demo_locality_list.go` package.

- the inter-region simulated latency map is moved from `cli/demo.go`
  to a new file `democluster/region_latencies.go`.

Release note: None

knz committed:
cli/sql: extract the dispatch code of the shell into a new file

This ensures there are no cobra dependencies remaining inside the
interactive shell.

Release note: None

knz committed:
cli/sql: split internal tests into a new file

"internal" tests are those that are interested in the internals of the
SQL shell. The remaining tests in `sql_test.go` only use the external
CLI.

Release note: None

knz committed:
cli/sql: stop depending on global variables

This commit ensures that the interactive shell logic only uses
internal state and stops relying on global variables.

Release note: None

knz committed:
cli/sql.go: stop looking at `EmbeddedMode`

The `EmbeddedMode` boolean is not used in the SQL shell. The two
functions in `sql.go` that used it really belong to the `cliCtx` (and
incidentally are currently only used from `demo.go`). This commit
moves them away from `sql.go`.

Release note: None

knz committed:
democluster: pull the interactive part of the API into separate pkg

In the longer term, we will want the ability to instantiate a SQL
interactive shell without the need to link in all the server code.

This requires the ability to compile the SQL shell without a
dependency on the `democluster` package.

Towards this goal, this commit extracts the part of the
`democluster.DemoCluster` interface API used in the SQL shell into a
separate package.

Release note: None

knz committed:
cli/sql: move `welcomeMsg` out of `sql.go`

since it's not used there.

Release note: None

knz committed:
cli/sql: further encapsulate the shell

This patch segregates the configuration variables that are only
visible from within the interactive shell away from `sqlCtx` into a
new `sqlInternalContext`.

Additionally, it stops the shell from looking directly at `os.Stdout`
and `stderr` and instead makes the out/err streams configurable upon
invocation.

Release note: None

knz committed:
cli: move the SQL shell to its own `clisqlshell` package

At last! The SQL shell is now independent from global state
and from the CockroachDB server machinery.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67462)
<!-- Reviewable:end -->

knz committed:
cli: new format 'rawhtml' to auto-gen the settings table

Previously, the cluster setting markdown table was generated with a
`gen` sub-command that was trying to be too clever with the HTML
generator. This was making it complicated to move the table formatting
code to a separate package.

This commit simplifies the code by introducing a new output format
'rawhtml' (raw, as in, it does not escape HTML special characters
inside table cells).

Since this format is only used internally to auto-generate this table,
the help texts leave it undocumented.

Release note: None

otan commented on pkg/cli/democluster/api/api.go:
should the package name match `api`?

knz committed:
clisqlclient: document the table output formats

Release note: None

knz commented on pkg/cli/democluster/api/api.go:
The package name is the prefix that's going to be used in calls by client packages.
If the reader sees `api.XXX` it does not say "which API it is".
With `democlusterapi.XXX` it's clearer where XXX comes from.

Meanwhile, I chose the `api` directory name because the immediate parent directory is `democluster` so it's kind of clear *when looking at the file system path* which API we're looking at.

We have many other packages where the package name does not match the import path - that's a standard Go feature.
